lib/puppet/application/resource.rb

```diff
@@ -14,13 +15,17 @@ def preinit
   option("--to_yaml","-y")
 
   option("--types", "-t") do |arg|
-    types = []
-    Puppet::Type.loadall
-    Puppet::Type.eachtype do |t|
-      next if t.name == :component
-      types << t.name.to_s
+    env = Puppet.lookup(:environments).get(Puppet[:environment]) || create_default_environment!
+
+    Puppet.override(:current_environment => env, :loaders => Puppet::Pops::Loaders.new(env)) do
+      types = []
+      Puppet::Type.loadall
```

Comment:
You can call `Puppet::Type.loadall(env)` to make it more explicit that we're loading types from a specific environment. If one isn't specified, then `loadall` falls back to using the context (so no behavior change), but I think passing an explicit environment is clearer.

Reply:
@joshcooper it looks like `Puppet::Type.loadall(env)` calls `loadall` from [`MetaType`](https://github.com/puppetlabs/puppet/blob/b7c9dd387a23718ba552d07e16f32533ff343b5a/lib/puppet/metatype/manager.rb#L52-L54) which doesn't allow arguments.
We could potentially call `Puppet::Type.typeloader.loadall(env)`?

Reply:
ah right, either way is fine.